api/lib/infrastructure/repositories/organization-learner-repository.js

```diff
@@ -168,30 +157,45 @@ module.exports = {
-    );
-    const students = await studentRepository.findReconciledStudentsByNationalStudentId(
-      _.compact(nationalStudentIdsFromFile),
-      domainTransaction
-    );
+  async _reconcileOrganizationLearners(studentsToImport, allOrganizationLearnersInSameOrganization, domainTransaction) {
+    const nationalStudentIdsFromFile = studentsToImport
+      .map((organizationLearnerData) => organizationLearnerData.nationalStudentId)
+      .filter(Boolean);
+    const organizationLearnersWithSameNationalStudentIdsAsImported =
+      await studentRepository.findReconciledStudentsByNationalStudentId(nationalStudentIdsFromFile, domainTransaction);
 
-    _.each(students, (student) => {
-      const alreadyReconciledOrganizationLearner = _.find(organizationLearnersToImport, {
-        userId: student.account.userId,
-      });
+    organizationLearnersWithSameNationalStudentIdsAsImported.forEach((organizationLearner) => {
+      const alreadyReconciledStudentToImport = studentsToImport.find(
+        (studentToImport) => studentToImport.userId === organizationLearner.account.userId
+      );
 
-      if (alreadyReconciledOrganizationLearner) {
-        alreadyReconciledOrganizationLearner.userId = null;
-      } else if (_canReconcile(existingOrganizationLearners, student)) {
-        const organizationLearner = _.find(organizationLearnersToImport, {
-          nationalStudentId: student.nationalStudentId,
-        });
-        organizationLearner.userId = student.account.userId;
+      if (alreadyReconciledStudentToImport) {
+        alreadyReconciledStudentToImport.userId = null;
+        return;
+      }
+
+      const organizationLearnerWithSameUserId = allOrganizationLearnersInSameOrganization.find(
+        (organizationLearnerInSameOrganization) => {
+          return organizationLearnerInSameOrganization.userId === organizationLearner.account.userId;
+        }
+      );
+      const organizationLearnerHasSameUserIdAndNationalStudentId =
+        organizationLearnerWithSameUserId?.nationalStudentId === organizationLearner.nationalStudentId;
+      const isOrganizationLearnerReconciled = organizationLearnerWithSameUserId != null;
+
+      if (!isOrganizationLearnerReconciled || organizationLearnerHasSameUserIdAndNationalStudentId) {
+        const studentToImport = studentsToImport.find(
+          (studentToImport) => studentToImport.nationalStudentId === organizationLearner.nationalStudentId
+        );
+        const isFromSameOrganization = studentToImport.organizationId === organizationLearner.account.organizationId;
+        const isFromDifferentOrganizationWithSameBirthday =
+          !isFromSameOrganization && studentToImport.birthdate === organizationLearner.account.birthdate;
+        const shouldStudentToImportBeReconciled = isFromSameOrganization || isFromDifferentOrganizationWithSameBirthday;
+
+        if (shouldStudentToImportBeReconciled) studentToImport.userId = organizationLearner.account.userId;
```

Comment:
je trouve que remettre tout ça dans la méthode complexifie la lecture, c'est dommage...

Reply:
```suggestion
      if (alreadyReconciledStudentToImport) {
        alreadyReconciledStudentToImport.userId = null;
        return;
      }
      
      if (_shouldStudentToImportBeReconciled(params)) studentToImport.userId = organizationLearner.account.userId;
```
En mettant tout dans la méthode `_shouldStudentToImportBeReconciled` c'est plus lisible